### PR TITLE
feat: Add flag checking ALL or ANY to waiting lists

### PR DESCRIPTION
**Pull request by AxonC**
This PR introduces a new `flags_check` field to the waiting lists table.
This facilitates a waiting list looking for either all of its flags to have been met
or any of the ones which have been defined.

By default, waiting lists will be the 'all' type in the database, but the type must be manually selected
when creating the waiting list via Nova. 

*Path to production:* This should be ready to deploy. The migration will add the column and default
all existing waiting lists to use the 'all' type. This is in line with current behaviour. The lists
which use the 'any' type e.g. the C1 waiting list can be edited via Nova and the list will automatically
adjust.

### CODE REVIEW

File: database/migrations/2021_05_22_092019_add_any_or_all_flag_to_waiting_list_flags.php [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+{
+    /**
+     * Run the migrations.
+     *
+     * @return void
+     */
+    public function up()
+    {
+        Schema::table('training_waiting_list', function (Blueprint $table) {
+            $table->string('flags_check')->default(WaitingList::ALL_FLAGS);
```
Comment (atoff) [inline database/migrations/2021_05_22_092019_add_any_or_all_flag_to_waiting_list_flags.php:18]: For table neatness?

```suggestion
            $table->string('flags_check')->default(WaitingList::ALL_FLAGS)->after('department');
```

File: tests/Unit/Training/WaitingList/WaitingListAccountTest.php [hunk 2]
```diff
@@ -2,6 +2,7 @@
 
 namespace Tests\Unit\Training\WaitingList;
 
+use App\Models\Atc\Endorsement\Condition;
```
Comment (atoff) [inline tests/Unit/Training/WaitingList/WaitingListAccountTest.php:5]: I don't see this import used in this file?

**Review by atoff — COMMENTED** [review]
Looks good but just some minor suggestions

File: app/Models/Training/WaitingList/WaitingListAccount.php [hunk 3]
```diff
@@ -147,14 +147,16 @@ public function allFlagsChecker()
...
-        $this->flags()->each(function ($model) use (&$checked) {
-            if (! $model->pivot->value) {
-                $checked = false;
-
-                return false;
-            }
-        });
+        if ($this->waitingList->flags_check == WaitingList::ALL_FLAGS) {
+            // iterate through each of the flags to see if they are true. If a false flag is detected, stop iterating.
+            $checked = $this->flags->every(function ($model) {
```
[the remark this answers is not in the document]
Comment (AxonC) [reply app/Models/Training/WaitingList/WaitingListAccount.php]: I was 50/50 on this but you convinced me :) 

https://github.com/VATSIM-UK/core/pull/2184/commits/42a85f4f6ccbe274081298697cb3376cf87bf907

File: tests/Unit/Training/WaitingList/WaitingListAccountTest.php (hunk 2 above)
Comment (AxonC) [reply tests/Unit/Training/WaitingList/WaitingListAccountTest.php]: https://github.com/VATSIM-UK/core/pull/2184/commits/a02fbed20ffe8c985c40cba1fce6b0937aacd75d

File: database/migrations/2021_05_22_092019_add_any_or_all_flag_to_waiting_list_flags.php (hunk 1 above)
Comment (AxonC) [reply database/migrations/2021_05_22_092019_add_any_or_all_flag_to_waiting_list_flags.php]: Good call https://github.com/VATSIM-UK/core/pull/2184/commits/fbc9863153adea260a53833b01d379479c122eed

File: app/Models/Training/WaitingList/WaitingListAccount.php [hunk 4]
```diff
@@ -145,18 +145,18 @@ public function getAtcHourCheckAttribute()
...
-
-                return false;
-            }
-        });
+        if ($this->waitingList->flags_check == WaitingList::ALL_FLAGS) {
+            // iterate through each of the flags to see if they are true. If a false flag is detected, stop iterating.
+            return $this->flags->every(function ($model) {
+                return $model->pivot->value;
+            });
+        } elseif ($this->waitingList->flags_check == WaitingList::ANY_FLAGS && $this->flags->count() > 0) {
```
Comment (atoff) [inline app/Models/Training/WaitingList/WaitingListAccount.php:153]: I'll leave it up to you if you want to remove the else here 😜

**Review by atoff — APPROVED** [review]

**Review by CalumTowers — APPROVED** [review]
👍 LGTM
Happy to leave this for you to merge when you're ready and configure in production as required.

**Pull request merged**